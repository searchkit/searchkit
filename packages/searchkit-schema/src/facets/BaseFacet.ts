import QueryManager, { MixedFilter } from '../core/QueryManager'

export interface FacetOptions {
  size: number
  query: string
}

export interface BaseFacet {
  excludeOwnFilters: boolean
  getIdentifier(): string
  getLabel(): string
  getAggregation(overrides: FacetOptions, queryManager: QueryManager): any
  getFilters(filters: Array<MixedFilter>): any
  transformResponse(response: any, queryManager: QueryManager): any
  getSelectedFilter(filterSet: any): any
}
