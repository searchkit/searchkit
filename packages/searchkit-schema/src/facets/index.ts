export { default as RefinementSelectFacet } from './RefinementSelectFacet'
export { default as RangeFacet } from './RangeFacet'
export { default as DateRangeFacet } from './DateRangeFacet'
export { default as HierarchicalMenuFacet } from './HierarchicalMenuFacet'
export * from './BaseFacet'
export * from './VisibilityRules'
