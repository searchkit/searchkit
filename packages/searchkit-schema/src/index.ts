export * from './resolvers'
export * from './query'
export * from './facets'
export * from './filters'

export { default as SearchkitSchema, SearchkitSchemaConfig } from './schema'
