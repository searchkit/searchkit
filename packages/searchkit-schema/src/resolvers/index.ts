export { default as HitsResolver } from './HitsResolver'
export { default as SearchkitResolver } from './ResultsResolver'
export { default as FacetsResolver } from './FacetsResolver'
export { default as SummaryResolver } from './SummaryResolver'
export { default as FacetResolver } from './FacetResolver'
