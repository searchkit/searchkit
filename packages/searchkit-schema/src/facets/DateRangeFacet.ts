import _ from 'lodash'
import { DateRangeFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface DateRangeFacetConfig {
  identifier: string
  field: string
  label: string
  size?: number
  display?: 'DateRange' | string
}

class DateRangeFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: DateRangeFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getIdentifier() {
    return this.config.identifier
  }

  getFilters(filters: Array<DateRangeFilter>) {
    const rangeFilter: { gte?: string; lte?: string } = {}
    if (!_.isUndefined(filters[0].dateMin)) rangeFilter.gte = filters[0].dateMin
    if (!_.isUndefined(filters[0].dateMax)) rangeFilter.lte = filters[0].dateMax
    return { range: { [this.config.field]: rangeFilter } }
  }

  getAggregation() {
    return {}
  }

  getSelectedFilter(filterSet) {
    return {
      type: 'DateRangeSelectedFilter',
      id: `${this.getIdentifier()}_${filterSet.dateMin}_${filterSet.dateMax}`,
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      dateMin: filterSet.dateMin,
      dateMax: filterSet.dateMax,
      display: this.config.display || 'DateRangeFacet'
    }
  }

  transformResponse() {
    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      type: 'DateRangeFacet',
      display: this.config.display || 'DateRangeFacet',
      entries: []
    }
  }
}

export default DateRangeFacet
