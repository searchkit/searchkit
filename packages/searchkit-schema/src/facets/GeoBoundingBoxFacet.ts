import { DateRangeFilter, GeoBoundingBoxFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface GeoBoundingBoxFacetConfig {
  identifier: string
  field: string
  label: string
  display?: 'DateRange' | string
}

class GeoBoundingBoxFacet implements BaseFacet {
  public excludeOwnFilters = false

  constructor(public config: GeoBoundingBoxFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getIdentifier() {
    return this.config.identifier
  }

  getFilters(filters: Array<GeoBoundingBoxFilter>) {
    return {
      geo_bounding_box: {
        [this.config.field]: {
          top_left: filters[0].geoBoundingBox.topLeft,
          bottom_right: filters[0].geoBoundingBox.bottomRight
        }
      }
    }
  }

  getAggregation() {
    return {}
  }

  getSelectedFilter(filterSet) {
    const c = filterSet.geoBoundingBox
    return {
      type: 'GeoBoundingBoxSelectedFilter',
      id: `${this.getIdentifier()}_${c.topLeft.lat}_${c.topLeft.lon}_${c.bottomRight.lat}_${
        c.bottomRight.lon
      }`,
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      topLeft: c.topLeft,
      bottomRight: c.bottomRight,
      display: null
    }
  }

  transformResponse() {
    return null
  }
}

export default GeoBoundingBoxFacet
