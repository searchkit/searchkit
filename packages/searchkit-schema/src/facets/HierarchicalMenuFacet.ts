import _ from 'lodash'
import QueryManager, { HierarchicalValueFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface HierarchicalMenuFacetConfig {
  identifier: string
  fields: Array<string>
  label: string
  display?: 'HierarchicalMenu' | string
}

class HierarchicalMenuFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: HierarchicalMenuFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getIdentifier() {
    return this.config.identifier
  }

  getFilters(filters: Array<HierarchicalValueFilter>) {
    return null
  }

  getAggregation(overrides, queryManager: QueryManager) {
    const appliedFilters =  queryManager.getFiltersById(this.config.identifier) as Array<HierarchicalValueFilter> || []

    const levelAggs = this.config.fields.reduce((aggs, field, index) => {
      const level = index + 1
      const parentFilters = appliedFilters.filter((f) => f.level < level)
      const getAggs = (parentFilters.length === (level - 1))
      if (getAggs) {
        return {
          ...aggs,
          [`lvl_${level}`]: {
            "filter": parentFilters.length === 0 ? { match_all: {} } : {
              "bool": {
                "must": parentFilters.map((f) => {
                  return {
                    "term": {
                      [this.config.fields[f.level-1]]: {
                        "value": f.value
                      }
                    }
                  }
                })
              },
            },
            aggs: {
              "aggs": {
                "terms": {
                  "field": this.config.fields[level-1]
                }
              }
            }
          }
        }
      } else {
        return aggs
      }
    }, {})

    return {
      [this.getIdentifier()]: {
        "filter": {
          "match_all": {}
        },
        "aggs": {
          ...levelAggs
        }
      }
    }
  }

  getSelectedFilter(filterSet) {
    return {
      type: 'HierarchicalMenuFilter',
      id: `${this.getIdentifier()}`,
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      value: filterSet.value,
      level: filterSet.level,
      display: this.config.display || 'HierarchicalMenuFacet'
    }
  }

  transformResponse(response, queryManager: QueryManager) {
    const appliedFilters = queryManager.getFiltersById(this.config.identifier) as Array<HierarchicalValueFilter> || []
    const x = (level) => {
      if (response[`lvl_${level}`]) {
        const levelFilter = appliedFilters.find((f) => f.level === level)
        return response[`lvl_${level}`].aggs.buckets.map((bucket) => {
          return {
            label: bucket.key,
            count: bucket.doc_count,
            id: `${this.getIdentifier()}_${bucket.key}_${level}`,
            entries: levelFilter?.value === bucket.key ? x(level + 1) : null
          }
        })
      } else {
        return null
      }
    }

    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      type: 'HierarchicalMenuFacet',
      display: this.config.display || 'HierarchicalMenuFacet',
      entries: x(1)
    }
  }
}

export default HierarchicalMenuFacet
