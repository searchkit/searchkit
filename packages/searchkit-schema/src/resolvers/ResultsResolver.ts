import { RequestBody } from '@elastic/elasticsearch/lib/Transport'
import QueryManager, { MixedFilter } from '../core/QueryManager'
import SearchkitRequest from '../core/SearchkitRequest'
import BaseQuery from '../query/BaseQuery'
import { BaseFacet } from '../facets/BaseFacet'
import { BaseFilter } from '../filters/BaseFilter'

export interface SortingOption {
  id: string
  label: string
  field: any
  defaultOption?: boolean
}

export interface SearchkitConfig {
  host: string
  index: string
  sortOptions?: SortingOption[]
  hits: {
    fields: string[]
  }
  query?: BaseQuery
  facets?: Array<BaseFacet>
  filters?: Array<BaseFilter>
  postProcessRequest?: (body: RequestBody) => RequestBody
}

export interface ResultsResolverParameters {
  filters: Array<MixedFilter>
  query: string
}

export default async (parent, parameters, ctx, info) => {
  try {
    const returnTypeName = info.returnType.name
    const config = ctx.searchkit.configs[returnTypeName]
    const skConfig = {
      sortOptions: [],
      ...config
    }
    const baseFilters = ctx.searchkit.baseFilters[returnTypeName]
      ? ctx.searchkit.baseFilters[returnTypeName](parent, parameters, ctx, info)
      : []
    const queryManager = new QueryManager(parameters.filters, parameters.query)
    const skRequest = new SearchkitRequest(queryManager, skConfig, baseFilters)

    return {
      searchkit: {
        skRequest: skRequest,
        queryManager: queryManager,
        config: skConfig,
        hitType: ctx.searchkit.hitTypeMappings[returnTypeName]
      }
    }
  } catch (e) {
    console.log(e)
  }
}
