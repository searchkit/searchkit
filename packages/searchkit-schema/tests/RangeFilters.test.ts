import nock from 'nock'
import { SearchkitConfig } from '../src/resolvers/ResultsResolver'
import { MultiMatchQuery } from '../src'
import { DateRangeFacet, RangeFacet } from '../src/facets'
import { setupTestServer, callQuery } from './support/helper'
import HitsMock from './__mock-data__/FacetsResolver/results.json'

describe('Range Facets', () => {
  describe('should return as expected', () => {
    const runQuery = async (gql) => {
      const response = await callQuery({ gql })
      return response
    }

    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new RangeFacet({
          field: 'imdbRating',
          label: 'IMDB Rating',
          identifier: 'imdbRating',
          range: {
            min: 0,
            max: 100,
            interval: 10
          }
        }),
        new DateRangeFacet({
          field: 'released',
          label: 'Released',
          identifier: 'released'
        })
      ]
    }

    it('numeric range optional max', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(query: "", filters: [
            { identifier: "imdbRating", min: 0 }
          ]) {
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "aggs": Object {},
              "from": 0,
              "post_filter": Array [
                Object {
                  "range": Object {
                    "imdbRating": Object {
                      "gte": 0,
                    },
                  },
                },
              ],
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('date range optional max', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(query: "", filters: [
            { identifier: "released", dateMin: "2012-12-18T00:00:00.000Z" }
          ]) {
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "aggs": Object {},
              "from": 0,
              "post_filter": Array [
                Object {
                  "range": Object {
                    "released": Object {
                      "gte": "2012-12-18T00:00:00.000Z",
                    },
                  },
                },
              ],
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })
  })
})
