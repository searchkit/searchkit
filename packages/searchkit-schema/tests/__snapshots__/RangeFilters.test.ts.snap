// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range Facets should return as expected date range optional max 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "id": "tt0468569",
        },
        Object {
          "id": "tt0110912",
        },
        Object {
          "id": "tt0167260",
        },
        Object {
          "id": "tt0114369",
        },
        Object {
          "id": "tt0073486",
        },
        Object {
          "id": "tt0172495",
        },
        Object {
          "id": "tt1520211",
        },
        Object {
          "id": "tt0338013",
        },
        Object {
          "id": "tt0078788",
        },
        Object {
          "id": "tt0180093",
        },
      ],
    },
  },
}
`;

exports[`Range Facets should return as expected numeric range optional max 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "id": "tt0468569",
        },
        Object {
          "id": "tt0110912",
        },
        Object {
          "id": "tt0167260",
        },
        Object {
          "id": "tt0114369",
        },
        Object {
          "id": "tt0073486",
        },
        Object {
          "id": "tt0172495",
        },
        Object {
          "id": "tt1520211",
        },
        Object {
          "id": "tt0338013",
        },
        Object {
          "id": "tt0078788",
        },
        Object {
          "id": "tt0180093",
        },
      ],
    },
  },
}
`;
